src/html/XHPHelpers.php

```diff
@@ -187,4 +134,54 @@ public function requireUniqueID(): string {
     }
   }
 
+  protected function getAttributesThatAppendValuesOnTransfer(): ImmSet<string> {
+    return ImmSet { 'class' };
+  }
+
+  final public function transferAttributesToRenderedRoot(
+    :x:composable-element $root,
+  ): void {
+    if (:xhp::$ENABLE_VALIDATION && $root instanceof :x:element) {
+      if (!($root instanceof HasXHPHelpers)) {
+        throw new XHPClassException(
+          $this,
+          'render() must return an object using the XHPHelpers trait.'
+        );
+      }
+
+      $rootID = $root->getAttribute('id') ?: null;
+      $thisID = $this->getAttribute('id') ?: null;
+
+      if ($rootID && $thisID && $rootID != $thisID) {
+        throw new XHPException(
+          'ID Collision. '.(:xhp::class2element(self::class)).' has an ID '.
+          'of "'.$thisID.'" but it renders into a(n) '.
+          (:xhp::class2element(get_class($root))).
+          ' which has an ID of "'.$rootID.'". The latter will get '.
+          'overwritten (most often unexpectedly). If you are intending for '.
+          'this behavior consider calling $this->removeAttribute(\'id\') '.
+          'before returning your node from compose().'
+        );
+      }
+    }
+    assert($root instanceof HasXHPHelpers);
+
+    $attributes = $this->getAttributes();
+
+    // We want to append classes to the root node, instead of replace them,
+    // so do this attribute manually and then remove it.
+    foreach ($this->getAttributesThatAppendValuesOnTransfer() as $attr) {
+      $rootValue = (string) $root->getAttribute($attr);
+      $thisValue = array_key_exists($attr, $attributes)
+        ? (string) $attributes[$attr]
+        : '';
+      if ($rootValue !== '' && $thisValue !== '') {
```

Comment:
I //think// this needs to be `||` since otherwise the root's class could be overwritten by an empty string, right?